kv: Test partial partition between two nodes.

andrewbaptist opened the pull request:
Partial partitions are difficult to deal with. This test creates partial partitions with the leases for key ranges placed on the nodes that are being disconnected and observes the latency behavior.

Release note: None
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95394)
<!-- Reviewable:end -->

andrewbaptist commented:
This test passes today on master, but fails if the commit for #87244 is reverted. I ran it a few times and it passed every time, but I have a small concern that it will uncover additional issues that we haven't covered yet.

erikgrinaker commented on pkg/cmd/roachtest/tests/failover.go:
```diff
@@ -32,6 +32,13 @@ import (
 
 func registerFailover(r registry.Registry) {
+	r.Add(registry.TestSpec{
+		Name:    "failover/partial-disconnect",
```
We should have tests for multiple kinds of partial partitions. I enumerated the ones I think are important in [the doc](https://docs.google.com/document/d/134eVs69fVBhP-01bxfrdjPxz--drNjOQ8AI-7Z3UD6U/edit). Let's call this e.g. `failover/partial/lease-liveness` so that we can add additional kinds later.

Should we test both blackhole and TCP reset failure cases? I think blackhole is probably sufficient since that's what we're most likely to see in the wild, and the TCP reset case is strictly easier to handle, but we could consider adding it for completeness.

erikgrinaker commented on pkg/cmd/roachtest/tests/failover.go:
```diff
@@ -70,6 +77,74 @@ func registerFailover(r registry.Registry) {
+// 5 nodes fully connected. Break the connection between a pair of nodes 4 and 5
+// while running a workload against nodes 1 through 3. Before each disconnect,
+// move all the leases to nodes 4 and 5 in a different pattern.
+func runDisconnect(ctx context.Context, t test.Test, c cluster.Cluster) {
```
Let's try to be more specific about the scenario we're testing here, to have tigher controls on the failure mode and required solution (which also avoids unintended flake). We specifically want to test a partial partition between a user range leaseholder and the liveness range leaseholder -- no more, no less. I suggest we place ranges as follows:

- System ranges: n1,n2,n3
- Liveness range: n1,n2,n4 with lease on n4
- User ranges: n2,n3,n5 with leases on n5
- SQL gateways: n1,n2,n3

This is almost what we have here already, but we should use zone configurations to ensure we e.g. don't end up with the meta range on n4 and that starts messing with things (we'll unfortunately also want to move ranges manually to avoid having to wait for the replicate queue). This also ensures that `WaitFor3XReplication` actually waits for all upreplication, since it won't wait for e.g. 4x or 5x replication which could cause this to flake.

erikgrinaker commented on pkg/cmd/roachtest/tests/failover.go:
```diff
@@ -70,6 +77,74 @@ func registerFailover(r registry.Registry) {
+		// All the system ranges will be on all the nodes, so they will all move,
+		// plus many of the non-system ranges.
+		for i := 2; i <= 7; i++ {
+			time.Sleep(30 * time.Second)
```
Let's use 1 minute intervals here. I found that several failure modes can take more than 30 seconds to resolve, and we want to capture these in the benchmark results as well.

Let's also use a similar timer approach as in e.g. `runFailoverNonSystem`. Tickers instead of sleeps ensure that the time taken to e.g. relocate leases or ranges doesn't skew the overall runtime, which could otherwise cause the workload to stop before we're done with partitions. And the random wait ensures we start the partition at different offsets into the lease interval, since a fixed periodicity will otherwise skew the results towards artificially low pMax latencies.

erikgrinaker commented on pkg/cmd/roachtest/tests/failover.go:
```diff
@@ -70,6 +77,74 @@ func registerFailover(r registry.Registry) {
+		// Try different combinations of the ranges between the partitioned nodes.
+		// All the system ranges will be on all the nodes, so they will all move,
+		// plus many of the non-system ranges.
+		for i := 2; i <= 7; i++ {
```
Why 2-7? Let's do 9 iterations to match the other benchmarks, I found we needed that many to get a reasonably reliable pMax given the inherent randomness here.

erikgrinaker commented on pkg/cmd/roachtest/tests/failover.go:
```diff
@@ -70,6 +77,74 @@ func registerFailover(r registry.Registry) {
+			qps := measureQPS(ctx, t, conn, 5*time.Second)
+			t.Status("Node 1 QPS after waiting 30 seconds is: ", qps)
+			// There is some rounding error with QPS, so verify that it is > 1
+			require.True(t, qps > 1.0)
```
We shouldn't assert anything here. These tests are intended to be more like benchmarks rather than pass/fail tests, since we [graph them with roachperf](https://roachperf.crdb.dev/?filter=&view=failover%2Fnon-system%2Fcrash&tab=gce) and want to track availability improvements over time. Failing the test won't record a data point for that run (I think).

erikgrinaker commented on pkg/cmd/roachtest/tests/failover.go:
```diff
@@ -640,6 +715,26 @@ func makeFailer(
+				if reconnect {
+					c.Run(ctx, c.Node(nodes[n]), `sudo iptables -F`)
+				} else {
+					c.Run(ctx, c.Node(nodes[n]), `sudo iptables -A INPUT -s `+ips[ip]+` -j DROP`)
```
This is sort of an asymmetric partition too. It'll effectively prevent TCP connections from working since we block all return traffic, but for completeness we should add a corresponding `OUTPUT` rule too.

erikgrinaker commented on pkg/cmd/roachtest/tests/failover.go:
```diff
@@ -640,6 +715,26 @@ func makeFailer(
+		for ip := 0; ip < len(ips); ip++ {
+			if n != ip {
+				if reconnect {
+					c.Run(ctx, c.Node(nodes[n]), `sudo iptables -F`)
```
We'll need to run this cleanup on all nodes whenever the test fails too, since the cluster can be reused for other roachtests which would otherwise get weird failures because networking is broken. A `defer` should do it.

erikgrinaker commented on pkg/cmd/roachtest/tests/failover.go:
```diff
@@ -70,6 +77,74 @@ func registerFailover(r registry.Registry) {
+			require.True(t, qps > 1.0)
+
+			t.Status("recovering n4 and n5")
+			Disconnect(t, c, ctx, []int{4, 5}, true)
```
nit: consider a separate function for this -- `Disconnect` is the exact opposite of what we're doing here. :)

andrewbaptist commented:
A refactored the code a little to make it easier to run more tests. Also I allowed it to run in local mode (it just doesn't create the real partitions). This is useful for faster iteration without running on real clusters.

Commit:
kv: Test partial partition between two nodes.

Partial partitions are difficult to deal with. This test creates
partial partitions with the leases for key ranges placed on the nodes
that are being disconnected and observes the latency behavior.

Release note: None

erikgrinaker commented on pkg/cmd/roachtest/tests/failover.go:
```diff
@@ -66,6 +75,97 @@ func registerFailover(r registry.Registry) {
+	randTimer := time.After(randutil.RandDuration(rng, max))
+	// Randomly sleep up to the lease renewal interval, to vary the time
+	// between the last lease renewal and the failure. We start the timer
+	// before the range relocation above to run them concurrently.
```
nit: This comment belongs at the call site, not in the function, and it's no longer accurate either.

erikgrinaker commented on pkg/cmd/roachtest/tests/failover.go:
```diff
@@ -140,9 +229,7 @@ func runFailoverNonSystem(
-	_, err = conn.ExecContext(ctx, `ALTER DATABASE kv CONFIGURE ZONE USING `+
-		`num_replicas = 3, constraints = '[-node1, -node2, -node3]'`)
-	require.NoError(t, err)
+	constrainConfig(t, ctx, conn, `DATABASE kv`, 3, []int{1, 2, 3}, 0)
```
nit: Not sure I think this is an improvement. It's harder to read and understand, and less flexible, just to avoid copying a few lines of code.

erikgrinaker commented on pkg/cmd/roachtest/tests/failover.go:
```diff
@@ -624,14 +712,20 @@ type blackholeFailer struct {
+func (f *blackholeFailer) Ready(_ context.Context, _ cluster.Monitor) {}
 
 func (f *blackholeFailer) Cleanup(ctx context.Context) {
+	if f.c.IsLocal() {
```
We'll need to do this across all of the failers. We should also log a warning about it, so people don't start wondering why it isn't doing anything.

andrewbaptist reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95394)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @herkolategan, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/failover.go` line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r1-36:-NMeD7EK5uOwQWx6gWxZ:biyspmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc7e6b365a4391c9e1f8dae4ad4d1a1cdb9bba/pkg/cmd/roachtest/tests/failover.go#L36)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

We should have tests for multiple kinds of partial partitions. I enumerated the ones I think are important in [the doc](https://docs.google.com/document/d/134eVs69fVBhP-01bxfrdjPxz--drNjOQ8AI-7Z3UD6U/edit). Let's call this e.g. `failover/partial/lease-liveness` so that we can add additional kinds later.

Should we test both blackhole and TCP reset failure cases? I think blackhole is probably sufficient since that's what we're most likely to see in the wild, and the TCP reset case is strictly easier to handle, but we could consider adding it for completeness.
</blockquote></details>

Done (rename). I agree this is the "harder" case, and was trying to avoid adding too many new tests related to this. I want to add a more complete "randomized" testing to this once we have all these tests passing, but for now I wasn't going to add a new one.
___
*[`pkg/cmd/roachtest/tests/failover.go` line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r1-83:-NMeDsGm2zTdOhp_vbn3:b3k6vlq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc7e6b365a4391c9e1f8dae4ad4d1a1cdb9bba/pkg/cmd/roachtest/tests/failover.go#L83)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Let's try to be more specific about the scenario we're testing here, to have tighter controls on the failure mode and required solution (which also avoids unintended flake). We specifically want to test a partial partition between a user range leaseholder and the liveness range leaseholder -- no more, no less. I suggest we place ranges as follows:

- System ranges: n1,n2,n3
- Liveness range: n1,n2,n4 with lease on n4
- User ranges: n2,n3,n5 with leases on n5
- SQL gateways: n1,n2,n3

This is almost what we have here already, but we should use zone configurations to ensure we e.g. don't end up with the meta range on n4 and that starts messing with things (we'll unfortunately also want to move ranges manually to avoid having to wait for the replicate queue). This also ensures that `WaitFor3XReplication` actually waits for all upreplication, since it won't wait for e.g. 4x or 5x replication which could cause this to flake.
</blockquote></details>

Thanks! I didn't realize why you were doing some of the things in the previous test (like changing the system zones to 3 replicas) - but this makes a lot more sense. Updated the test
___
*[`pkg/cmd/roachtest/tests/failover.go` line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r1-124:-NMeE5vf3NFCRQXKU9nY:blwzcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc7e6b365a4391c9e1f8dae4ad4d1a1cdb9bba/pkg/cmd/roachtest/tests/failover.go#L124)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Why 2-7? Let's do 9 iterations to match the other benchmarks, I found we needed that many to get a reasonably reliable pMax given the inherent randomness here.
</blockquote></details>

Done - This was a carry-over when I was initially distributing the replicas based on %i. I changed that but didn't fix this.
___
*[`pkg/cmd/roachtest/tests/failover.go` line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r1-125:-NMeF--ZCvzJal5zwL6S:b-nawcu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc7e6b365a4391c9e1f8dae4ad4d1a1cdb9bba/pkg/cmd/roachtest/tests/failover.go#L125)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Let's use 1 minute intervals here. I found that several failure modes can take more than 30 seconds to resolve, and we want to capture these in the benchmark results as well.

Let's also use a similar timer approach as in e.g. `runFailoverNonSystem`. Tickers instead of sleeps ensure that the time taken to e.g. relocate leases or ranges doesn't skew the overall runtime, which could otherwise cause the workload to stop before we're done with partitions. And the random wait ensures we start the partition at different offsets into the lease interval, since a fixed periodicity will otherwise skew the results towards artificially low pMax latencies.
</blockquote></details>

Done - thanks for the tip.
___
*[`pkg/cmd/roachtest/tests/failover.go` line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r1-138:-NMerL9AAXVq64bA3o8G:b-gn2hks) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc7e6b365a4391c9e1f8dae4ad4d1a1cdb9bba/pkg/cmd/roachtest/tests/failover.go#L138)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

We shouldn't assert anything here. These tests are intended to be more like benchmarks rather than pass/fail tests, since we [graph them with roachperf](https://roachperf.crdb.dev/?filter=&view=failover%2Fnon-system%2Fcrash&tab=gce) and want to track availability improvements over time. Failing the test won't record a data point for that run (I think).
</blockquote></details>

Removed assert - however it might be better to have these pass/fail as well as tracking availability, I worry that if we don't have some rudimentary failure checking we won't detect breakage if no-one is looking at these. We should only add asserts for ones that currently pass (like this one and a couple others) - not the ones that are currently failing.
___
*[`pkg/cmd/roachtest/tests/failover.go` line 141 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r1-141:-NMerJNc04iqS9zh-Llv:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc7e6b365a4391c9e1f8dae4ad4d1a1cdb9bba/pkg/cmd/roachtest/tests/failover.go#L141)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: consider a separate function for this -- `Disconnect` is the exact opposite of what we're doing here. :)
</blockquote></details>

Done
___
*[`pkg/cmd/roachtest/tests/failover.go` line 729 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r1-729:-NMerI_ME34WqenpQUFE:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc7e6b365a4391c9e1f8dae4ad4d1a1cdb9bba/pkg/cmd/roachtest/tests/failover.go#L729)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

We'll need to run this cleanup on all nodes whenever the test fails too, since the cluster can be reused for other roachtests which would otherwise get weird failures because networking is broken. A `defer` should do it.
</blockquote></details>

Done
___
*[`pkg/cmd/roachtest/tests/failover.go` line 731 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r1-731:-NMerHrq7w-V0JhWtIxA:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc7e6b365a4391c9e1f8dae4ad4d1a1cdb9bba/pkg/cmd/roachtest/tests/failover.go#L731)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

This is sort of an asymmetric partition too. It'll effectively prevent TCP connections from working since we block all return traffic, but for completeness we should add a corresponding `OUTPUT` rule too.
</blockquote></details>

Done
___
*[`pkg/cmd/roachtest/tests/failover.go` line 82 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r3-82:-NP3ozhy4a3vjOInJ5qG:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/11839dcf349976baf19c0719a820c16b053f4111/pkg/cmd/roachtest/tests/failover.go#L82)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: This comment belongs at the call site, not in the function, and it's no longer accurate either.
</blockquote></details>

Done
___
*[`pkg/cmd/roachtest/tests/failover.go` line 232 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r3-232:-NP3p3Ky3IIfCdSU_FZ-:b-xmware) ([raw file](https://github.com/cockroachdb/cockroach/blob/11839dcf349976baf19c0719a820c16b053f4111/pkg/cmd/roachtest/tests/failover.go#L232)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: Not sure I think this is an improvement. It's harder to read and understand, and less flexible, just to avoid copying a few lines of code.
</blockquote></details>

I'm not sure it really improved things either. There were a lot of repeated calls, and it was easy to make a mistake before, but I understand the concern about reducing flexibility. Leaving for now. 
___
*[`pkg/cmd/roachtest/tests/failover.go` line 719 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r3-719:-NP3qqiw5HnP_sEJNXfP:b-210zf6) ([raw file](https://github.com/cockroachdb/cockroach/blob/11839dcf349976baf19c0719a820c16b053f4111/pkg/cmd/roachtest/tests/failover.go#L719)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

We'll need to do this across all of the failers. We should also log a warning about it, so people don't start wondering why it isn't doing anything.
</blockquote></details>

I added logging for this and extended to all the places where it doesn't work on local. Some tests do still apply on local.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (approved):
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95394)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @herkolategan, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/failover.go` line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r1-138:-NPqydoQCjluLNPWWGec:b-lbfln5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc7e6b365a4391c9e1f8dae4ad4d1a1cdb9bba/pkg/cmd/roachtest/tests/failover.go#L138)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Removed assert - however it might be better to have these pass/fail as well as tracking availability, I worry that if we don't have some rudimentary failure checking we won't detect breakage if no-one is looking at these. We should only add asserts for ones that currently pass (like this one and a couple others) - not the ones that are currently failing.
</blockquote></details>

The same applies for all of our benchmarks. If we e.g. see a 50% performance drop on kv0, we don't alert on that, but rely on periodic inspection and test-eng's regression detector to alert us to regressions. This could definitely be improved, but the way to do that is to add support for alerting in roachperf such that we get both data recording and alerting (preferably by just using Grafana instead of roachperf, which already supports that).
___
*[`pkg/cmd/roachtest/tests/failover.go` line 719 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r3-719:-NPr-6UW1pMevjpcAVIF:bk3oo38) ([raw file](https://github.com/cockroachdb/cockroach/blob/11839dcf349976baf19c0719a820c16b053f4111/pkg/cmd/roachtest/tests/failover.go#L719)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

I added logging for this and extended to all the places where it doesn't work on local. Some tests do still apply on local.
</blockquote></details>

We'll need to add it in `diskStallFailer` too.
___
*[`pkg/cmd/roachtest/tests/failover.go` line 729 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NPqzH3247ezj_H5HuYs:-NPqzH3247ezj_H5HuYt:bzhqxdr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e138487a887796141ea35bf3bbc69b363f99626/pkg/cmd/roachtest/tests/failover.go#L729)):*
> ```Go
> func (f *blackholeFailer) Fail(ctx context.Context, nodeID int) {
> 	if f.c.IsLocal() {
> 		f.t.Status("skipping disconnect cleanup on local cluster")
> ```

nit: this isn't cleanup, it's the actual failure.


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
TFTR

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-:-NPrHVSx7zPC_oRrgzjh:bb73zsl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @herkolategan, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/failover.go` line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r1-138:-NPrDLUa4i5q4OCp32oB:bhgzg7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fbc7e6b365a4391c9e1f8dae4ad4d1a1cdb9bba/pkg/cmd/roachtest/tests/failover.go#L138)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

The same applies for all of our benchmarks. If we e.g. see a 50% performance drop on kv0, we don't alert on that, but rely on periodic inspection and test-eng's regression detector to alert us to regressions. This could definitely be improved, but the way to do that is to add support for alerting in roachperf such that we get both data recording and alerting (preferably by just using Grafana instead of roachperf, which already supports that).
</blockquote></details>

Yep - I agree - which is why it is removed. Longer term balancing the line between "detecting soon enough" and "not adding flakes" is tricky. I'd like to review at a bigger team meeting to see what we think is right as it seems a little unclear today. Ideally, we could create some check conditions without a high risk of them being flaky. 
___
*[`pkg/cmd/roachtest/tests/failover.go` line 719 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NM0VxojfuLXSbZRETXb-r3-719:-NPrDgC70pKPptIMIWBX:bez0lvr) ([raw file](https://github.com/cockroachdb/cockroach/blob/11839dcf349976baf19c0719a820c16b053f4111/pkg/cmd/roachtest/tests/failover.go#L719)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

We'll need to add it in `diskStallFailer` too.
</blockquote></details>

I added a TODO here. The disk stall failer already has some support for different environments, and this mode doesn't actually start at all. Without a larger change to `dmsetupDiskStaller` this was going to be difficult to do. I added a TODO to the test to add this support eventually but didn't want to try and add this to the PR. 
___
*[`pkg/cmd/roachtest/tests/failover.go` line 729 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95394#-NPqzH3247ezj_H5HuYs:-NPrDn7L5gF-XQn46Kzs:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e138487a887796141ea35bf3bbc69b363f99626/pkg/cmd/roachtest/tests/failover.go#L729)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: this isn't cleanup, it's the actual failure.
</blockquote></details>

Fixed


<!-- Sent from Reviewable.io -->

andrewbaptist commented:
bors r=erikgrinaker
